autoendpoint/src/server/routers/fcm/router.rs

```diff
@@ -0,0 +1,398 @@
+        self.metrics
+            .incr_with_tags("notification.bridge.error")
+            .with_tag("platform", "fcmv1")
+            .with_tag("reason", reason)
+            .send();
+    }
+}
+
+#[async_trait(?Send)]
+impl Router for FcmRouter {
+    async fn route_notification(&self, notification: &Notification) -> ApiResult<RouterResponse> {
+        debug!(
+            "Sending FCM notification to UAID {}",
+            notification.subscription.user.uaid
+        );
+        trace!("Notification = {:?}", notification);
+
+        let router_data = notification
+            .subscription
+            .user
+            .router_data
+            .as_ref()
+            .ok_or(FcmError::NoRegistrationToken)?;
+        let fcm_token = router_data
+            .get("token")
+            .and_then(Value::as_str)
+            .ok_or(FcmError::NoRegistrationToken)?;
+        let app_id = router_data
+            .get("app_id")
+            .and_then(Value::as_str)
+            .ok_or(FcmError::NoAppId)?;
+        let ttl = MAX_TTL.min(self.settings.ttl.max(notification.headers.ttl as usize));
+        let message_data = self.build_message_data(notification)?;
+
+        // Send the notification to FCM
+        let client = self.clients.get(app_id).ok_or(FcmError::InvalidAppId)?;
+        trace!("Sending message to FCM: {:?}", message_data);
+        if let Err(e) = client.send(message_data, fcm_token.to_string(), ttl).await {
+            return Err(self.handle_error(e));
+        }
+
+        // Sent successfully, update metrics and make response
+        trace!("FCM request was successful");
+        self.incr_success_metrics(notification);
+
+        Ok(RouterResponse::success(
+            self.endpoint_url
+                .join(&format!("/m/{}", notification.message_id))
+                .expect("Message ID is not URL-safe")
+                .to_string(),
+            notification.headers.ttl as usize,
+        ))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::error::ApiErrorKind;
+    use crate::server::extractors::notification::Notification;
+    use crate::server::extractors::notification_headers::NotificationHeaders;
+    use crate::server::extractors::routers::RouterType;
+    use crate::server::extractors::subscription::Subscription;
+    use crate::server::routers::fcm::client::tests::{
+        make_service_file, mock_fcm_endpoint_builder, mock_token_endpoint, PROJECT_ID,
+    };
+    use crate::server::routers::fcm::error::FcmError;
+    use crate::server::routers::fcm::router::FcmRouter;
+    use crate::server::routers::RouterError;
+    use crate::server::routers::{Router, RouterResponse};
+    use crate::server::FcmSettings;
+    use autopush_common::db::DynamoDbUser;
+    use cadence::StatsdClient;
+    use std::collections::HashMap;
+    use std::path::PathBuf;
+    use url::Url;
+    use uuid::Uuid;
+
+    const FCM_TOKEN: &str = "test-token";
+    const CHANNEL_ID: &str = "4530d3a6-13f9-4639-87f9-2ff731824f34";
```

Comment:
Traditionally, I've used UUIDs that start with something like "DEADBEEF". This lets me scan and purge the database if the test leaks. 

(And it may not be your code that leaks. Someone later on could add something, use the testing value and that winds up corrupting things.)